woocommerce/woocommerce-ios
# Products & Product Variation: Allow alternative type for purchasable

Closes: #7510 


## Description

Some plugins alter the field value from a boolean to an integer. This causes issues in loading products in the app. As suggested in #7510, we can attempt to offer an alternative decoding if the expected one fails.

### Changes:

* In `Networking/Product` - update decoding of purchasable to accept both Bool and an alternative integer value.
* In `Networking/ProductVariation` - update decoding of purchasable to accept both Bool and an alternative integer value.
* In `Networking.KeyedDecodingContainer+Woo` - add integer support to `alternativeTypes`

## Testing instructions

According to p4a5px-2PC-p2, we found out the [YITH WooCommerce Gift Cards](https://wordpress.org/plugins/yith-woocommerce-gift-cards/) plugin is returning `0` or `1` values for the `purchasable` field. This plugin will be used for this test.

1. Install the [YITH WooCommerce Gift Cards](https://wordpress.org/plugins/yith-woocommerce-gift-cards/) extension on your store
2. Create a gift card at WP Admin > YITH > Gift Cards. Enter an amount (e.g., 10) and any current balance (> 0).
3. Create a new product at WP Admin > Products > Add New.
4. Set the `Product data` to `Gift card` and add any Gift card amounts.
5. Check the product fields via WC REST API using the product ID (e.g., `../wp-json/wc/v3/products/212`). The `purchasable` field should return 1. (Note: if you set the current balance to 0 in step 2, `purchasable` will be 0)
6. In the app, open the Products tab
7. All products should be shown

## Screenshots


**Before**

<img src="https://user-images.githubusercontent.com/40906847/212647670-e50dda07-11e3-44bc-9b74-8e7067bb079c.png" width=300 />

**After**

<img src="https://user-images.githubusercontent.com/40906847/212647140-5787c9d5-83a5-4670-95ad-d28fb9c02a4f.png" width=300 />

---
